### DISCUSSION TITLE
[Language Server] Support '@' keyword and expressions for COSMIC C Compilers

### DISCUSSION THREAD

**Discussion by github0null** [microsoft/vscode-cpptools#10830] [Ideas]
Hello, Everyone !

I've been using this plugin recently to help me write embedded programs for some microcontrollers (stm8...)

And i'm using the cosmic compiler: https://www.cosmicsoftware.com/index.php

And i'm using VSCode + cpptools as the code editor.

Most of the time this plugin works perfectly, but it won't be able to auto-complete code or trigger a hover when it encounters some special syntax.

I realized that this was because the plug-in could not understand non-standard syntax ('@' is a invalid token)

```c

/*	USART section
 */
volatile char USART_SR      @0x4a30;	/* USART Status Register */
volatile char USART_DR      @0x4a31;	/* USART Data Register */
volatile char USART_BRR1    @0x4a32;	/* USART Baud Rate Register 1 */
volatile char USART_BRR2    @0x4a33;	/* USART Baud Rate Register 2 */
volatile char USART_CR1     @0x4a34;	/* USART Control Register 1 */
volatile char USART_CR2     @0x4a35;	/* USART Control Register 2 */
volatile char USART_CR3     @0x4a36;	/* USART Control Register 3 */
volatile char USART_CR4     @0x4a37;	/* USART Control Register 4 */
volatile char USART_CR5     @0x4a38;	/* USART Control Register 5 */
volatile char USART_GTR     @0x4a39;	/* USART Guard Time Register */
volatile char USART_PSCR    @0x4a3a;	/* USART Prescaler Register */

/*	Send a char to the SCI.
 */
@near void outch(char c)
	{
	while (!(USART_SR & TRDE))	/* wait for READY */
		;
	USART_DR = c;			/* send it */
	}

/*	Character reception routine.
 *	This routine is called on interrupt.
 *	It puts the received char in the buffer.
 */
@interrupt void recept(void)
	{
	USART_SR;			/* clear interrupt */
	*ptecr++ = USART_DR;		/* get the char */
	if (ptecr >= &buffer[SIZE])	/* put it in buffer */
		ptecr = buffer;
	}

```

Is there any way to make cpptools blind to this syntax, just as it is to c comments ?

Because this syntax like a gcc attributes.

Or is it possible to support this syntax? If it's not too complicated, I can try to implement it myself

Or can you add support for this syntax ?

This is COSMIC C language Guide, most of the syntax is c99, with a few extensions:

[COSMIC-CLanguage.pdf](https://github.com/microsoft/vscode-cpptools/files/11238777/COSMIC-CLanguage.pdf)

**Comment by sean-mcmanus** [comment]
The only option is to use some `#define` like
```cpp
#if __INTELLISENSE__
#define COSMIC(x)
#else
#define COSMIC(x) x
#endif
```
and wrap all the uses in that macro.

You could file a feature request to add support for that, but it's unlikely that we would add it soon. It would require closed source changes to our IntelliSense parser.

**Reply by github0null** [reply]
Ok, thanks for your answer !
